seed/php-sdk/imdb/src/Core/SerializableType.php

```diff
@@ -0,0 +1,386 @@
+        return json_encode($arrayData, JSON_THROW_ON_ERROR);
+    }
+
+    /**
+     * Serializes the object to an array for JSON encoding.
+     *
+     * @return array
+     */
+    public function toArray(): array
+    {
+        $result = [];
+        $reflectionClass = new \ReflectionClass($this);
+
+        foreach ($reflectionClass->getProperties() as $property) {
+            $jsonKey = self::getJsonKey($property);
+            $value = $property->getValue($this);
+
+            // Handle DateTime properties
+            $dateTypeAttr = $property->getAttributes(DateType::class)[0] ?? null;
+            if ($dateTypeAttr && $value instanceof DateTime) {
+                $dateType = $dateTypeAttr->newInstance()->type;
+                $value = ($dateType === DateType::TYPE_DATE)
+                    ? $value->format('Y-m-d')
+                    : $value->format(DateTime::RFC3339);
+            }
+
+            // Handle arrays with type annotations
+            $arrayTypeAttr = $property->getAttributes(ArrayType::class)[0] ?? null;
+            if ($arrayTypeAttr && is_array($value)) {
+                $arrayType = $arrayTypeAttr->newInstance()->type;
+                $value = self::serializeGenericArray($value, $arrayType);
+            }
+
+            if ($value !== null) {
+                $result[$jsonKey] = $value;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Serializes the given array based on the type annotation.
+     *
+     * @param array $data The array to be serialized.
+     * @param array $type The type definition from the annotation.
+     * @return array The serialized array.
+     */
+    private static function serializeGenericArray(array $data, array $type): array
+    {
+        return self::isMapType($type)
+            ? self::serializeMap($data, $type)
+            : self::serializeList($data, $type);
+    }
+
+    /**
+     * Serializes a value based on the type definition.
+     *
+     * @param mixed $data The value to serialize.
+     * @param mixed $type The type definition.
+     * @return mixed The serialized value.
+     */
+    private static function serializeValue($data, $type)
+    {
+        if ($type instanceof Union) {
+            foreach ($type->types as $unionType) {
+                try {
+                    return self::serializeSingleValue($data, $unionType);
+                } catch (Exception $e) {
+                    continue;
+                }
+            }
+            throw new \InvalidArgumentException("Cannot serialize value with any of the union types.");
+        }
+
+        return self::serializeSingleValue($data, $type);
+    }
+
+    /**
+     * Serializes a single value based on its type.
+     *
+     * @param mixed $data The value to serialize.
+     * @param string $type The expected type.
+     * @return mixed The serialized value.
+     */
+    private static function serializeSingleValue($data, $type)
```

Comment:
nit: Should we use type declarations in the signature, too?
```suggestion
    private static function serializeSingleValue(mixed $data, string $type): mixed
```